### PR TITLE
Java docs (II)

### PR DESCRIPTION
**Pull request by VisheshVadhera**
Added docs for the following classes/interfaces:

- CacheControl
- ApolloInterceptor
- ApolloCacheInterceptor
- ApolloParseInterceptor
- ApolloServerInterceptor
- ApolloInterceptorChain
- RealApolloInterceptorChain

More docs on the way!

### CODE REVIEW

File: apollo-runtime/src/main/java/com/apollographql/apollo/interceptor/ApolloInterceptor.java [hunk 1]
```diff
@@ -11,20 +11,68 @@
...
+   * @param operation  the GraphQL Operation object contained within the outgoing request.
+   * @param chain      the ApolloInterceptorChain object containing the next set of interceptors.
+   * @param dispatcher the dispatcher which dispatches the non blocking operations on the request/response.
+   * @param callBack   the Callback which will handle the interceptor's response or failure exception.
+   */
   void interceptAsync(@Nonnull Operation operation, @Nonnull ApolloInterceptorChain chain,
       @Nonnull ExecutorService dispatcher, @Nonnull CallBack callBack);
 
+  /**
+   * Disposes off the resources which are no longer required.
```
Comment (brianPlummer) [inline apollo-runtime/src/main/java/com/apollographql/apollo/interceptor/ApolloInterceptor.java]: of*

File: apollo-runtime/src/main/java/com/apollographql/apollo/cache/normalized/CacheControl.java [hunk 2]
```diff
@@ -1,8 +1,33 @@
...
+   * Signals the apollo client to <b>only</b> fetch the data from the normalized cache. If it's not present in
+   * the normalized cache or if an exception occurs while trying to fetch it from the normalized cache, an empty {@link
+   * com.apollographql.apollo.api.Response} is sent back with the {@link com.apollographql.apollo.api.Operation} info
+   * wrapped inside.
+   */
   CACHE_ONLY,
+  /**
+   * Signals the apollo client to first fetch the data from the network. If network request fails, then the
+   * data is fetched from the normalized cache. If the data is not present in the normalized cache, then the
+   * exception leading to network request failure is rethrown.
```
Comment (wdziemia) [inline apollo-runtime/src/main/java/com/apollographql/apollo/cache/normalized/CacheControl.java]: Last sentence is slightly confusing, maybe something like:

>  then the exception which led to network request failure is rethrown.

File: apollo-runtime/src/main/java/com/apollographql/apollo/interceptor/ApolloInterceptor.java [hunk 3]
```diff
@@ -11,20 +11,68 @@
 
 import javax.annotation.Nonnull;
 
+/**
+ * ApolloInterceptor is responsible for observing and modifying the requests going out and and the corresponding
```
Comment (wdziemia) [inline apollo-runtime/src/main/java/com/apollographql/apollo/interceptor/ApolloInterceptor.java]: `and` is duplicated

File: apollo-runtime/src/main/java/com/apollographql/apollo/interceptor/ApolloInterceptor.java (hunk 1 above)
Comment (VisheshVadhera) [inline apollo-runtime/src/main/java/com/apollographql/apollo/interceptor/ApolloInterceptor.java]: Nice catch!

File: apollo-runtime/src/main/java/com/apollographql/apollo/interceptor/ApolloInterceptor.java [hunk 4]
```diff
@@ -11,20 +11,68 @@
...
+   * @param callBack   the Callback which will handle the interceptor's response or failure exception.
+   */
   void interceptAsync(@Nonnull Operation operation, @Nonnull ApolloInterceptorChain chain,
       @Nonnull ExecutorService dispatcher, @Nonnull CallBack callBack);
 
+  /**
+   * Disposes off the resources which are no longer required.
+   *
+   * <p>A UseCase for this method call would be when an {@link com.apollographql.apollo.ApolloCall} needs to be
+   * cancelled and resources need to be disposed off. </p>
```
Comment (wdziemia) [inline apollo-runtime/src/main/java/com/apollographql/apollo/interceptor/ApolloInterceptor.java]: of*

File: apollo-runtime/src/main/java/com/apollographql/apollo/interceptor/ApolloInterceptorChain.java [hunk 5]
```diff
@@ -6,10 +6,32 @@
 
 import javax.annotation.Nonnull;
 
+/**
+ * ApolloInterceptorChain is responsible for building the entire {@link ApolloInterceptorChain} chain.
+ */
 public interface ApolloInterceptorChain {
+
+  /**
+   * Passes the control over to the next {@link ApolloInterceptor} in the responsibility chain and blocks till the
```
Comment (wdziemia) [inline apollo-runtime/src/main/java/com/apollographql/apollo/interceptor/ApolloInterceptorChain.java]: until*

File: apollo-runtime/src/main/java/com/apollographql/apollo/interceptor/ApolloInterceptorChain.java [hunk 6]
```diff
@@ -6,10 +6,32 @@
...
+   * as this is a non blocking call. In order to receive the results back, pass in a callback which will handle the
+   * received response or error.
+   *
+   * @param dispatcher the {@link ExecutorService} which dispatches the calls asynchronously.
+   * @param callBack   the callback which will handle the response or a failure exception.
+   */
   void proceedAsync(@Nonnull ExecutorService dispatcher, @Nonnull ApolloInterceptor.CallBack callBack);
 
+  /**
+   * Disposes off the resources which are no longer required.
```
Comment (wdziemia) [inline apollo-runtime/src/main/java/com/apollographql/apollo/interceptor/ApolloInterceptorChain.java]: of*

File: apollo-runtime/src/main/java/com/apollographql/apollo/cache/normalized/CacheControl.java [hunk 7]
```diff
@@ -1,8 +1,33 @@
 package com.apollographql.apollo.cache.normalized;
 
+/**
+ * CacheControl class represents the cache controlling strategies for reading GraphQL responses back from the
+ * normalized cache when a network request is made.
```
Comment (sav007) [inline apollo-runtime/src/main/java/com/apollographql/apollo/cache/normalized/CacheControl.java]: We should omit this `when a network request is made.` as for instance cache only we don't make network call at all.

File: apollo-runtime/src/main/java/com/apollographql/apollo/internal/interceptor/ApolloServerInterceptor.java [hunk 8]
```diff
@@ -27,6 +27,9 @@
 import okhttp3.Response;
 import okio.Buffer;
 
+/**
+ * ApolloServerInterceptor is responsible for making the network calls to the server.
```
Comment (sav007) [inline apollo-runtime/src/main/java/com/apollographql/apollo/internal/interceptor/ApolloServerInterceptor.java]: We should probably add that this is the terminal interceptor (must be the last in order) as it doesn't call proceed on the chain.

File: apollo-runtime/src/main/java/com/apollographql/apollo/cache/normalized/CacheControl.java (hunk 7 above)
Comment (VisheshVadhera) [inline apollo-runtime/src/main/java/com/apollographql/apollo/cache/normalized/CacheControl.java]: Right

File: apollo-runtime/src/main/java/com/apollographql/apollo/internal/interceptor/ApolloServerInterceptor.java (hunk 8 above)
Comment (VisheshVadhera) [inline apollo-runtime/src/main/java/com/apollographql/apollo/internal/interceptor/ApolloServerInterceptor.java]: Done

**Comment by VisheshVadhera** [comment]
Have made the final set of changes. If everything looks fine, we can go ahead and merge it.

**Pull request merged**